### PR TITLE
WIP Enhance CSV Capabilities

### LINKED ISSUE #34: Not able to read CSV with multiple empty (duplicate column names)
Not able to read csv with multiple empty column names.
Expected behaviour : rename columns with numerica suffix such as empty1, empty2

example:
```csv
name, lastname , , , address
jay, patel, 12,123,12,3 na na
```

Errror
```bash
ReadCsv: Duplicate columns detected: [                       ]
```

### PR DESCRIPTION
**Pull request by kp2401075**
Read CSVs with Empty column names
Read CSVs with Duplicate Column names.

This is to fix #34

### CODE REVIEW

**Comment by kp2401075** [comment]
@tobgu Can you please review this and suggest any corrections.

Thanks,

**Comment by tobgu** [comment]
Thanks for the PR! I've given this some thought and have some ideas/suggestions that you may want to consider.

With the current solution we address two issues in one package (empty columns and duplicates) which is sort of fine but the user cannot control how this is done in any way (empty names are replaced with "column<colIx>" while other duplicates get the name "<name><colIx>"). The new config variable doesn't really provide a hint of how this is done. It simply states that we'll rename duplicates.

I have a suggestion: we move the details of how the rename and duplicate handling is done into a "policy function" and let this policy function be part of the `Config` object instead. The "policy function" could have the following signature:
`func(colIndex int, colName string) (newName string)` and we could set a default implementation of it that would simply be the identity function:
```go
func defaultPolicy(_ int, colName string) string {
   return colName
}
```

You can then implement more or less the code you've written in a new policy function which is constructed providing the prefix to use for empty columns as a string:

```go
// This would allow a user of the library to implement any policy they want.
func ColumnNamePolicy(func(colIndex int, colName string) string) ConfigFunc {
   // Return a function that sets policy func on config object
}

func AddColumnIndexToDuplicateColumns(emptyColNamePrefix string) ConfigFunc {
  policyFunc := func(colIx int, colName string) string {
    // Your existing code goes here
  }
  return ColumnNamePolicy(policyFunc)
}
```

In `ReadCSV` there would simply be a loop somewhere near the top to apply the policy function to each header index and name and replace the names read or passed in with those produced by the policy function.

There are obviously a few moving parts involved in the solution proposed above and there may be simpler ways to achieve the same thing. Please let me know your thoughts. Also, I realize I may have made a bad job trying to convey my ideas above. Please ask for clarifications in that case. I can provide some more code snippets for example.

Also, please add tests verifying the important parts of any new code that you add!

**Comment by kp2401075** [comment]
Hi @tobgu,

I think There is a simpler way to achieve this, I know I did a dumb thing and implemented 2 things in to 1.

but we can break that in to 2 config options.
1. For empty column names
2. For duplicate column names


Something like below:

and we can think of better name for config options.

```go
//For reading  CSV
type CSVConfig struct {
	EmptyNull        bool
	IgnoreEmptyLines bool
	Delimiter        byte
	Types            map[string]types.DataType
	EnumVals         map[string][]string
	RowCountHint     int
	Headers          []string
	AddColumnIndexToDuplicateColumns bool
	AddColumnIndexToEmptyColumns bool
}

```
we can give examples in Documentation for explaining what each option will do to user's Input. This should make it straight forward for user.

If you think solution I proposed is not a good fit then I can try and implement your proposal from https://github.com/tobgu/qframe/pull/39#issuecomment-841874131

Have a think and let me know.

**Comment by tobgu** [comment]
Hi!

I think something like what you describe above is probably good enough and, as you say, simpler.

I would suggest changing `AddColumnIndexToEmptyColumns bool` to something like `MissingColumnNameAlias string` to give the caller the possibility to set the name of empty columns rather than hard coding it in the lib. The default could be "", which would lead to the current behaviour, eg. rejecting empty column names. To make this work with multiple empty column names `AddColumnIndexToDuplicateColumns` would also have to be set to true.

Code examples are always great to add and would be highly appreciated!

**Comment by kp2401075** [comment]
I've made appropriate changes & also added some test functions to test when building.

I've also added some Comments in the config functions to show what this option does to the user input.

when the documentation gets generated it should show up.

Can you please take another pass at it.

File: internal/io/csv.go [hunk 1]
```diff
@@ -98,6 +101,37 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+	}
+
+	if conf.AddColumnIndexToDuplicateColumns {
+
+		headerSet := strings.NewEmptyStringSet()
+		for count, h := range headers {
+
+			if headerSet.Contains(h) {
+				h = h + fmt.Sprint(count)
+			} else {
```
Comment (tobgu) [inline internal/io/csv.go:126]: I think this part potentially needs to be thought through a bit more to make it bullet proof wrt to duplicates. As is now you could construct header combinations that will fail the uniqueness constraints.
`COL,COL,COL1`, `COL,COL2,COL`

While probably not super common I'd prefer if we could handle this since its a library and you don't really know what users will use it for. We could come up with a naming scheme like <NAME><COL>.<ENUMERATOR> where the ENUMERATOR is just added in case there are conflicts and is incremented until it does not conflict with an already existing column name.

It would be great to add a couple of unit tests for these corner cases.

File: internal/io/csv.go [hunk 2]
```diff
@@ -98,6 +101,37 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+				headerSet.Add(h)
+			}
+			headers[count] = h
+
```
Comment (tobgu) [inline internal/io/csv.go:130]: Remove empty line

File: internal/io/csv.go [hunk 3]
```diff
@@ -98,6 +101,37 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+		}
+		// fmt.Println("Duplicates have been renamed")
```
Comment (tobgu) [inline internal/io/csv.go:132]: Remove comment

File: internal/io/csv.go [hunk 4]
```diff
@@ -116,17 +150,20 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
 		duplicates := make([]string, 0)
 		headerSet := strings.NewEmptyStringSet()
 		for _, h := range headers {
+
```
Comment (tobgu) [inline internal/io/csv.go:153]: Remove empty line

File: internal/io/csv.go [hunk 5]
```diff
@@ -116,17 +150,20 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
 			}
-		}
 
+		}
 		return nil, nil, qerrors.New("ReadCsv", "Duplicate columns detected: %v", duplicates)
+
 	}
 
 	return dataMap, headers, nil
+
```
Comment (tobgu) [inline internal/io/csv.go:166]: Remove empty line

File: qframe_test.go [hunk 6]
```diff
@@ -1373,6 +1373,56 @@ mon
 	})
 }
 
+func TestQFrame_ReadCSVMissingColumnName(t *testing.T) {
+	input := `,COL2
+a,1.5
+b,2.25
```
Comment (tobgu) [inline qframe_test.go:1379]: Perhaps reduce row count to one line since only one row is needed?

File: qframe_test.go [hunk 7]
```diff
@@ -1373,6 +1373,56 @@ mon
...
+	expectedIn := `col,COL2
+a,1.5
+b,2.25
+c,3.0`
+
+	out := qframe.ReadCSV(strings.NewReader(input), csv.MissingColumnNameAlias("col"))
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+
+	// Output:
```
Comment (tobgu) [inline qframe_test.go:1390]: Remove this comment or migrate this test to be an example test. See the examples in "examples_test.go". Those are part of the documentation and adding the happy/base cases there may be a good idea.

File: qframe_test.go [hunk 8]
```diff
@@ -1373,6 +1373,56 @@ mon
...
+
+	expectedIn := `COL,COL1
+a,1.5
+b,2.25
+c,3.0`
+
+	out := qframe.ReadCSV(strings.NewReader(input), csv.AddColumnIndexToDuplicateColumns(true))
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+	// Output:
```
Comment (tobgu) [inline qframe_test.go:1415]: See comment about examples in test above.

File: config/csv/config.go [hunk 9]
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
...
+
+func MissingColumnNameAlias(MissingColumnNameAlias string) ConfigFunc {
+	return func(c *Config) {
+		c.MissingColumnNameAlias = MissingColumnNameAlias
+	}
+}
+
+// AddColumnIndexToDuplicateColumns configures if column names have duplicates
+// column index will be added to the column name
+// COL,COL 			// COL0   COL1
```
Comment (tobgu) [inline config/csv/config.go:50]: I think we can skip the inline examples here and instead create proper example tests covering the usage.

File: config/csv/config.go [hunk 10]
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
 	}
 }
 
+// MissingColumnNameAliasString set configs to rename empty column name with given string
+//  ,COL2 			// col   COL2
```
Comment (tobgu) [inline config/csv/config.go:37]: Skip inline example here? See similar comment below.

File: config/csv/config.go [hunk 11]
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
 	}
 }
 
+// MissingColumnNameAliasString set configs to rename empty column name with given string
```
Comment (tobgu) [inline config/csv/config.go:36]: "MissingColumnNameAliasString..." -> "MissingColumnNameAlias sets the name to be used for empty columns..."
Perhaps add a comment here about the fact that you probably also want to set `AddColumnIndexToDuplicateColumns` if you expect to have multiple columns with missing names?

File: config/csv/config.go [hunk 12]
```diff
@@ -33,6 +33,31 @@ func EmptyNull(emptyNull bool) ConfigFunc {
...
+// b,2.25  				b    2.25
+// c,3.0`				c       3
+
+func MissingColumnNameAlias(MissingColumnNameAlias string) ConfigFunc {
+	return func(c *Config) {
+		c.MissingColumnNameAlias = MissingColumnNameAlias
+	}
+}
+
+// AddColumnIndexToDuplicateColumns configures if column names have duplicates
```
Comment (tobgu) [inline config/csv/config.go:48]: "... if duplicate column names should have the column index appended to the column name to resolve the conflict.", or similar?

File: internal/io/csv.go [hunk 13]
```diff
@@ -20,13 +21,15 @@ type bytePointer struct {
...
-	Headers          []string
+	EmptyNull                        bool
+	IgnoreEmptyLines                 bool
+	Delimiter                        byte
+	Types                            map[string]types.DataType
+	EnumVals                         map[string][]string
+	RowCountHint                     int
+	Headers                          []string
+	AddColumnIndexToDuplicateColumns bool
+	MissingColumnNameAlias           string ""
```
Comment (tobgu) [inline internal/io/csv.go:32]: What are the double quotes for?

**Review by tobgu — CHANGES_REQUESTED** [review]
Thanks! I've dropped a couple of comments/questions/suggestions that you may want to consider.

File: config/csv/config.go (hunk 9 above)
Comment (kp2401075) [reply config/csv/config.go:50]: I need help with this, What should be the function names in examples_test.go file. for examples.

File: internal/io/csv.go [hunk 14]
```diff
@@ -98,6 +101,27 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+			if name == "" {
+				headers[i] = conf.MissingColumnNameAlias
+			}
+		}
+	}
+
+	if conf.AddColumnIndexToDuplicateColumns {
+		headerSet := strings.NewEmptyStringSet()
+		for i, h := range headers {
+			if headerSet.Contains(h + fmt.Sprint(i)) {
```
Comment (tobgu) [inline internal/io/csv.go:115]: What would happen for ´COL1,COL,COL11´ (which is valid as is) for example?

I think you somehow need to take into consideration not only the headers that you've looped over but also those that come later since you don't want to create conflicts with existing column names. Perhaps a ´map[string]int´ mapping the column name to the first column position of the name  that you populate initially would be a better data structure for this than the stringset?

File: qframe_test.go [hunk 15]
```diff
@@ -1373,6 +1373,65 @@ mon
 	})
 }
 
+func TestQFrame_ReadCSVMissingColumnName(t *testing.T) {
```
Comment (tobgu) [inline qframe_test.go:1376]: These tests would probably lend themselves nicely to "table driven testing" if that's something you want to explore: https://github.com/golang/go/wiki/TableDrivenTests

File: internal/io/csv.go (hunk 14 above)
Comment (kp2401075) [reply internal/io/csv.go:115]: I knew this was going to be issue, I was thinking through something. might be able to test it next commit.

File: internal/io/csv.go [hunk 16]
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+				headers[i] = conf.MissingColumnNameAlias
+			}
+		}
+	}
+
+	if conf.AddColumnIndexToDuplicateColumns {
+		headersMap := make(map[string]int)
+		// Loop through all columsn and find the duplicates.
+		for i, h := range headers {
+			headersMap[h] = i
```
Comment (tobgu) [inline internal/io/csv.go:116]: I personally would prefer if this was the first index of the header rather than the last.
Eg. `COL,COL,COL` -> `COL,COL0,COL1`, right now it's `COL,COL,COL` -> `COL0,COL1,COL`.

File: internal/io/csv.go [hunk 17]
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+		}
+		// For go through all if there is duplicate down the line add index
+		for i, h := range headers {
+			index, ok := headersMap[h]
+			if ok && i != index {
+				counter := 0
```
Comment (tobgu) [inline internal/io/csv.go:122]: I like this approach better than the previous since it's more straight forward to reason about. I would like to change the name of the config variable though since the new approach has nothing to do with the column indicies.
Perhaps `RenameDuplicateColumns`?

File: internal/io/csv.go [hunk 18]
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+				for {
+					_, ok = headersMap[headers[i]+fmt.Sprint(counter)]
```
Comment (tobgu) [inline internal/io/csv.go:124]: Assign `headers[i]+fmt.Sprint(counter)` to a variable (`candidateName`?) since it's used multiple times?

Comment (tobgu) [inline internal/io/csv.go:124]: Inline and assign `_, ok =` in the if-expression to reduce the scope of the variable and avoid overwriting the existing `ok` variable?

File: internal/io/csv.go [hunk 19]
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
...
+	if conf.MissingColumnNameAlias != "" {
+		for i, name := range headers {
+			if name == "" {
+				headers[i] = conf.MissingColumnNameAlias
+			}
+		}
+	}
+
+	if conf.AddColumnIndexToDuplicateColumns {
+		headersMap := make(map[string]int)
```
Comment (tobgu) [inline internal/io/csv.go:113]: Consider making a new function for this code block taking headers and returning new headers since this is now becoming a quite significant piece of code.

File: internal/io/csv.go [hunk 20]
```diff
@@ -98,6 +101,40 @@ func ReadCSV(reader io.Reader, conf CSVConfig) (map[string]interface{}, []string
 		}
 	}
 
+	if conf.MissingColumnNameAlias != "" {
+		for i, name := range headers {
```
Comment (tobgu) [inline internal/io/csv.go:105]: Consider moving to its own function for symmetry with the below suggestion.

File: qframe_test.go [hunk 21]
```diff
@@ -1373,6 +1373,65 @@ mon
...
+	expectedIn := `COL0,COL
+a,1.5`
+
+	out := qframe.ReadCSV(strings.NewReader(input), csv.AddColumnIndexToDuplicateColumns(true), csv.MissingColumnNameAlias("COL"))
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+}
+
+func TestQFrame_ReadCSVDuplicateColumnNameWithColumnIndexAdded(t *testing.T) {
+	//Corner case 1
```
Comment (tobgu) [inline qframe_test.go:1412]: I'm not sure these are corner cases anymore with the new method.

File: qframe_test.go (hunk 15 above)
Comment (kp2401075) [reply qframe_test.go:1376]: I think I'm going to have to spend some more time getting familiar with TableDrivenTests.

Might do another PR for documentation seprately if that's ok.

File: config/csv/config.go [hunk 22]
```diff
@@ -33,6 +33,20 @@ func EmptyNull(emptyNull bool) ConfigFunc {
...
 
+// MissingColumnNameAlias sets the name to be used for empty columns name with given string
+func MissingColumnNameAlias(MissingColumnNameAlias string) ConfigFunc {
+	return func(c *Config) {
+		c.MissingColumnNameAlias = MissingColumnNameAlias
+	}
+}
+
+// AddColumnIndexToDuplicateColumns configures if duplicate column names should have the column index appended to the column name to resolve the conflict.
+func AddColumnIndexToDuplicateColumns(AddColumnIndexToDuplicateColumns bool) ConfigFunc {
```
Comment (tobgu) [inline config/csv/config.go:44]: As mentioned elsewhere please rename this configuration option now that the column indices are not part of the solution. `RenameDuplicateColumns`?

File: internal/io/csv.go [hunk 23]
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
 	}
 }
 
+// Handle DuplicateColumnNames
+func AddColumnIndexToDuplicateColumns(headers []string) []string {
+	headersMap := make(map[string]int)
+	// Loop through all columsn and find the duplicates.
```
Comment (tobgu) [inline internal/io/csv.go:168]: Consider rewriting this comment to better reflect why this is done or rename `headersMap` to something more telling and remove comment.

File: internal/io/csv.go [hunk 24]
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
...
+	for i, h := range headers {
+		_, ok := headersMap[h]
+		if !ok {
+			headersMap[h] = i
+		}
+	}
+	// For go through all if there is duplicate down the line add index
```
Comment (tobgu) [inline internal/io/csv.go:175]: Consider rewriting this comment to better reflect why this is done or removing it completely.

File: internal/io/csv.go [hunk 25]
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
 	}
 }
 
+// Handle DuplicateColumnNames
+func AddColumnIndexToDuplicateColumns(headers []string) []string {
```
Comment (tobgu) [inline internal/io/csv.go:166]: Make function module private by changing name to start with lower case.

File: internal/io/csv.go [hunk 26]
```diff
@@ -151,6 +162,48 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
...
+				}
+			}
+		}
+	}
+	return headers
+
+}
+
+// Handle Missing Columnnames
+func AddAliasToMissingColumnNames(headers []string, alias string) []string {
```
Comment (tobgu) [inline internal/io/csv.go:198]: Make function module private by changing name to start with lower case.

File: qframe_test.go (hunk 15 above)
Comment (tobgu) [reply qframe_test.go:1376]: If you have things that you know you'd like to do I'd prefer if we did it all in this PR to make it as complete. Otherwise I'm OK with the current state of affairs. Examples are great  but not a must.

File: qframe_test.go [hunk 27]
```diff
@@ -1373,6 +1373,41 @@ mon
...
+	expectedIn := `col,COL2
+a,1.5`
+
+	out := qframe.ReadCSV(strings.NewReader(input), csv.MissingColumnNameAlias("col"))
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+}
+
+func TestQFrame_ReadCSVDuplicateColumnName(t *testing.T) {
+	input := `COL,COL
```
Comment (tobgu) [inline qframe_test.go:1388]: Please update this input to something like `COL,COL,COL,KOL,KOL` to better capture the essence of the current behavior.

**Review by tobgu — CHANGES_REQUESTED** [review]
Great! I've added a couple of minor comments. Once those are fixed I think we're good to merge!

File: config/csv/config.go [hunk 28]
```diff
@@ -33,6 +33,20 @@ func EmptyNull(emptyNull bool) ConfigFunc {
...
 }
 
+// MissingColumnNameAlias sets the name to be used for empty columns name with given string
+func MissingColumnNameAlias(MissingColumnNameAlias string) ConfigFunc {
+	return func(c *Config) {
+		c.MissingColumnNameAlias = MissingColumnNameAlias
+	}
+}
+
+// AddColumnIndexToDuplicateColumns configures if duplicate column names should have the column index appended to the column name to resolve the conflict.
```
Comment (tobgu) [inline config/csv/config.go:43]: Update doc string to reflect new name and functionality

File: internal/io/csv.go [hunk 29]
```diff
@@ -151,6 +162,49 @@ func resizeColBytes(bytes [][]byte, currentRowCount, sizeHint int) {
 	}
 }
 
+//
```
Comment (tobgu) [inline internal/io/csv.go:165]: Remove empty comment

File: qframe_test.go [hunk 30]
```diff
@@ -1373,6 +1373,41 @@ mon
...
+a,1.5`
+	expectedIn := `COL,COL2
+a,1.5`
+
+	out := qframe.ReadCSV(strings.NewReader(input), csv.MissingColumnNameAlias("COL"))
+	expected := qframe.ReadCSV(strings.NewReader(expectedIn))
+	assertEquals(t, expected, out)
+}
+
+func TestQFrame_ReadCSVDuplicateColumnName(t *testing.T) {
```
Comment (tobgu) [inline qframe_test.go:1387]: This test does not assert anything meaningful ATM, there is an error in the header (a missing comma between col two and three) that cause the read method to fail. The resulting frame has `Err != nil`. Please fix this and also perhaps add a check that there is no error when reading the frame (`assertNotErr(t, out.Err)`) to be on the safe side.

**Review by tobgu — CHANGES_REQUESTED** [review]
Still some things that need fixing.

Once done please squash all commits into one.

Thanks!

**Comment by kp2401075** [comment]
@tobgu , I've tried to squash the commits. not sure it it worked correctly.

Please. review and see if it's really ready for merge.

File: examples_test.go [hunk 31]
```diff
@@ -92,13 +93,45 @@ c,3.0`
 	f := qframe.ReadCSV(strings.NewReader(input))
 	fmt.Println(f)
 	// Output:
-	// COL1(s) COL2(f)
+	// COL(s) COL0(f)
```
Comment (tobgu) [inline examples_test.go:96]: Not sure what happened here but I can take a look at it. Perhaps the examples need to be added to the automated tests.

**Review by tobgu — APPROVED** [review]
Great, thanks for the effort!

**Comment by tobgu** [comment]
Thanks a lot!
